html/editing/dnd/drop/events-contenteditable.tentative.html

```diff
@@ -0,0 +1,27 @@
+<!doctype html>
+<meta charset=utf-8>
+<title>Selection drag and drop: events for &lt;input></title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<style>
+[data-placeholder]:empty::before { content: attr(data-placeholder); }
+</style>
+<div><input id=a value="Drag me"></div>
+<div contenteditable=true id=b data-placeholder="...to here"></div>
+<script>
+setup({ explicit_timeout: true, single_test: true });
+const a = document.getElementById('a');
+const b = document.getElementById('b');
+const actualEvents = [];
+const expectedEvents = ['drop', 'beforeinput', 'textInput', 'input'];
+for (const eventType of expectedEvents) {
+  b.addEventListener(eventType, e => {
+    actualEvents.push(e.type);
+    if (actualEvents.length === expectedEvents.length) {
```

Comment:
This means that not assertion would be run if `textInput` is not implemented or dispatched.

How about to run the assertion with:

```
if (e.type == "input") {
  requestAnimationFrame(() =>
    requestAnimationFrame(() => assert_array_equals(...))
  });
}
```
These events should be dispatched by browsers immediately (synchronously). Therefore, checking it asynchronously enqueued from an event listener must work.

Reply:
And could you check whether `textInput` is not fired in the drag source editor but `beforeinput` and `input` should be fired.

Reply:
And also perhaps, there should be another set of tests to check events when the drag source and drop target is same editor.

Reply:
Done.

In WebKit and Chromium, `textInput` is fired on the target before the *source* `beforeinput` event. See http://software.hixie.ch/utilities/js/live-dom-viewer/saved/12422

But the test expects `textInput` after the target `beforeinput` event, which matches your implementation. Or, the test doesn't compare relative order between source and target. Maybe it should?

Reply:
> In WebKit and Chromium, `textInput` is fired on the target before the _source_ `beforeinput` event. See http://software.hixie.ch/utilities/js/live-dom-viewer/saved/12422

Hmm, that's odd to me because it does not match the event order when the drag source and drop target is same editable element.

> But the test expects `textInput` after the target `beforeinput` event, which matches your implementation.

Yeah, I believe that's reasonable. It's odd to swap the event order only in the DnD case. I guess that it was for making the `deleteByDrag` cancelable only with `textInput` before `beforeinput` era...

> Or, the test doesn't compare relative order between source and target. Maybe it should?

Well, it may be reasonable for now. Could you? I think you can do it only with changing the JS file to parse the `data-expexted-events` value.

Reply:
Yeah relative order is checked (see my previous comment above).